boltdesignsystem/bolt
# Add new Lazy Queue package to allow components to defer loading

## Jira
http://vjira2:8080/browse/BDS-2104

## Summary
Packages up the core work to help us _significantly_ boost our client-side web performance in the design system.

## Details
In a nutshell, this package allows any web component to register a custom element tag (or HTML tags) with a callback that runs once one of those tags is on the page.

To put it another way: this lets us have a very small JS bundle that loads on every page and defer loading almost all of the component-specific code in Bolt except for those elements that are originally present + dynamically load any additional code for new components that get added later on.

Beyond adding this new lazyQueue helper, this PR also minimally wires it up to the Video and Chip components to demonstrate actual usage + give us a few data points to start to measure against.

Internally this uses the [Idle Until Urgent pattern](https://philipwalton.com/articles/idle-until-urgent/) to prioritize the UI that's already there and more lazily handle any UI that _isn't_. 

## How to test
- Review code updates in this PR
- Confirm that the two components wired up here load up and work exactly as expected (meaning, when they don't exist on the page, the JS bundles for these never load and if they do end up getting added later on, the JS bundles get loaded dynamically)

Note: while this PR setups up the core functionality used under the hood, the most exciting changes (and largest perceived impact with this) is combining this work with the other related perf updates.
